text/0052-resource-importing-support.md

```diff
@@ -0,0 +1,268 @@
+specified file, or it will be passed the resource mapping directly from the
+previous phase.
+
+First, the sanity checks are performed and then the change set is created. By
+default, the change set is also executed (can be changed by passing
+`--no-execute` flag just like with regular deploy).
+
+#### Cannot import and create/change resources in the same changeset
+
+CloudFormation API doesn't allow any non-import changes in the template for
+import operation. This is a challenge for CDK, as:
+
+* `CDKMetadata` resource changes on each synthesis, so the import operation
+  never succeeds with the unmodified template. To resolve this, the proposed
+  solution restores CDKMetadata to the currently deployed version before
+  passing the template to ChangeSet creation step.
+* High-level constructs produce multiple underlying CloudFormation resources
+  and only some of those might need to be imported - the rest might need to
+  be created (e.g. `ec2.Vpc()` might require VPC being imported, but subnets
+  being created). Furthermore, CDK can't automatically determine what
+  resources to import and what to create. For this to behave correctly, the
+  solution will prompt user for each resource - whether it should be imported
+  or not. If not, the resource is removed from the template before creating
+  the change set. The remaining (not imported) resources can be created by
+  a subsequent `cdk deploy` command.
+
+#### An explicit DeletionPolicy must be set on imported resources
+
+CloudFormation import API requires all imported resources to have specified
+explicit DeletionPolicy. CDK satisfies this requirement for typical stateful
+resources (e.g. S3 buckets), but not _all_ resources (e.g. VPCs). The solution
+will inject a `DeletionPolicy` with a value of `Delete` (CloudFormation's
+default) in the template for all resources without `DeletionPolicy` set by CDK.
+
+This manipulation will only happen during the import operation and therefore it
+is not persistent. On the next deploy, the property is dropped. However, as
+`Delete` is the default value, semantically the template doesn't change.
+According to the testing, CloudFormation will allow dropping the explicit policy
+post-import.
+
+#### CloudFormation doesn't check the physical resource's configuration
+
+If a physical resource identified by the provided identifiers does not exist,
+the import operation (change set creation) will fail with clear-enough message.
+However, if the resource does exist, but it's actual configuration doesn't match
+the CloudFormation configuration, import operation succeeds and *not update*
+the resource, causing the stack to be drifted. This can be left to the user
+to handle (manually checking the config prior to the import or running drift
+detection after the import), but for convenience, this solution uses
+CloudControl API to look up the physical resource and validate its configuration
```

Comment:
This is worth calling out in the above section as well.

I assume this happens before doing the actual import?

Reply:
I wonder if running a drift detection afterwards might be good enough for a start?

The CCAPI will require us to do a JSON diff, and I'm not sure it's complete enough. Have you experimented with this?

Reply:
I have experimented with this and it works quite nicely. The greatest benefit is that CCAPI uses the same resource type nomenclature as CloudFormation so we can simply pass the type to the API call. There are two issues I am still looking at though: 

Firstly, I need to understand how the "import identifiers" of a resource relate to "primary" and "secondary" indentifiers that CCAPI uses. While for CFN import, explicit identifier _key_ has to be provided (e.g. `BucketName` for S3 Bucket, or `VpcId` for VPC), with CCAPI call you always provide `Identifier` key that has a value based on the type - it can be either primary or secondary identifier. My plan is to get all resource schemas from CFN resource catalog, extract primary/secondary identifiers info, then run get-template-summary for a template containing all resource types and do some analysis on that. It will be somewhat time consuming though, not sure when I can get to that.

Secondly, CCAPI returns _all_ resource properties, but some of them are not mandatory in CFN templates and if not specified, will be set to default values. So in order to do the comparison correctly, I would need to know the default values of all optional properties of all resources, didn't find a way to do that yet (in best-effort approach, we could only compare properties explicitly set in CFN).

I agree with your suggestion: If it is OK to get this merged without this extra validation, let's do it and either run drift detection automatically or display a post import message to tell user to run it. And leave the CCAPI extra validation for the next iteration.

Reply:
For now, I have updated the RFC to omit the CCAPI validations and I have added a "Possible future expansions" section with some thoughts on how to leverage CCAPI to provide a better user experience in future - perhaps beyond simple validations.